keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+    * [Garbage Collection](#garbage-collection)
+    * [Installation Status Reporting](#installation-status-reporting)
+    * [Health Check Pods](#health-check-pods)
+    * [Health Status Reporting.](#health-status-reporting)
+    * [Generation Observation](#generation-observation)
+    * [Adopting Existing Components](#adopting-existing-components)
+ * [kubectl](#kubectl)
+ * [UI Considerations](#ui-considerations)
+    * [Discovering Application Components and Dependencies](#discovering-application-components-and-dependencies)
+    * [Discovering an Object's Application](#discovering-an-objects-application)
+    * [Application Status](#application-status)
+ * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+ * [Planned](#planned)
+* [Drawbacks](#drawbacks)
+ * [Increased Memory, CPU, and Network Requirements](#increased-memory-cpu-and-network-requirements)
+ * [Increased API Server Storage Requirements](#increased-api-server-storage-requirements)
+* [Alternatives](#alternatives)
+ * [Use Labeling Conventions](#use-labeling-conventions)
+ * [Use Custom Resource Definitions](#use-custom-resource-definitions)
+ * [Application Type Aggregation](#application-type-aggregation)
+ * [Use the Garbage Collector Instead of an Application Controller](#use-the-garbage-collector-instead-of-an-application-controller)
+
+
+
+
+
+
+## Summary
+Kubernetes has many primitives for managing workloads (e.g. Pods, ReplicaSets, Deployments, DaemonSets and 
+StatefulSets), storage (e.g. PersistentVolumeClaims and PersistentVolumes), and networking (e.g. Services, 
+Headless Services, and Ingeresses). When these primitives are aggregated to provide a service to an end user or to 
+another system, the whole becomes something more than the individual parts. Instead of a set of loosely coupled 
+workloads and their corresponding storage and networking, we have an application. 
+
+Currently, there is no standard way for tools to discover the relationship between the components of an 
+application. Many tools use bespoke schemes (generally involving labeling and/or annotations). This has lead to a 
+proliferation on non-interoperable tools, a non-uniform experience for application developers and users, and an 
+inability of UI (User Interface) designers to surface information about applications to users.
+
+To address these issues, we propose to add an Application Kind as a first class citizen to the apps Group of the 
+Kuberentes API. This Kind can be used by tools to communicate that the applications they create are more than just 
+a loosely coupled set of API objects.
+
+The use of the Application object is strictly opt-in for tool developers. Tools and users that choose to use the 
+Application object in their manifests will benefit from interoperability, a uniform user experience, and first class 
+support by Kubernetes UIs.
+
```

Comment:
I left a ton of comments on this, mostly raising questions. But overall, I think this proposal does a _tremendous_ job of starting simple and trying hard to fend off unnecessary complexity. The implementation is actually broad enough that it does not encompass only applications (I left comments explaining this in context). I think it can be used to solve some interesting problems.

I would love to see this come to fruition (very soon) as a CRD/Controller, but I don't think it should be added to core straightaway.

Reply:
First, thank you for the thorough and thoughtful feedback.

Second, I do have some concerns about developing a CRD that we intend to eventually move to a core resource.

1. Conversion - If we release something as a CRD at beta level stability, how do our users upgrade their clusters to consume the Kuberentes resource that we eventually contribute? I could image writing a program that automates this process, but if the program does not work flawlessly, we introduce the potential to break users.
1. Fragmentation - If we eventually add a core object in place of a CRD, we can not expect all users of the CRD to migrate to the core object immediately. We will have to support both, and move the CRD through a deprecation cycle. During the interim, would we attempt to keep the  core resource compatible with the CRD so that a lossless conversion program will work? If so we would constrain the evolution of a core resource based on an external component.
1. Support - As corollary of (1) and (2), we would have to support the both the CRD and the core object concurrently for some duration. This adds an additional burden to both developers and administrators that support multiple clusters and different versions.

The only benefit to developing it as a CRD first would be to allow users to deploy it into live clusters at an alpha level of stability, and to allow its evolution to be independent of the core API. I'm more in favor of developing the right thing, in core, and promoting it as it matures.

Reply:
From a practical perspective, we already have to deal with multiple API versions and people running different version of Kubernetes and upgrades, don't we? For example, earlier today I was asking someone to switch from `extensions/v1beta1` for Deployments to `apps/v1beta2`. The reason for not going all the way up to `apps/v1` was that many people aren't using k8s 1.9, yet. It's not even available many places yet. Versions and support are messy. Helm has built in capabilities to detect versions and available APIs in chart templating to overcome this practical problem.

Tool authors and end users already have to deal with fragmentation, deprecation, and changes.

How would moving through API versions with a core object be different from doing it as a CRD and moving it in later?

Reply:
If there is general consensus that we eventually want to move it into core what is the benefit of starting it as a CRD as opposed to starting is an alpha resource and promoting it as stability matures?

Reply:
> If there is general consensus that we eventually want to move it into core

Is there general consensus? This seems like a distinct departure from the types of objects in core.

Reply:
@kow3ns Two things come to mind here, to add context to what you're seeing...

1. Maybe there isn't consensus. I previously thought there was enough interest and consensus but maybe there isn't. It may have to do with the nature of this object compared with others. So far the objects have been focused more on infrastructure and lower level features. This is a higher level organizational construct with an opinion. As a higher level organizational construct it's not one you'll find in something like AWS but instead in something like CloudFoundry. That's different enough that could cause some to pause on it being in core.
1. In SIG Architecture, if I remember correctly, the conversation has shifted from more features in core to a stable slow moving core with innovation in addons and plugins. If something proves to be popular and in wide enough use that it could be brought into the core.

Again, I'm just trying to add context and bring some points from differing conversations together.

Reply:
i havent read through the complete proposal. we at salesforce are doing a similar effort of making a application object using the CRD mechanism. I think putting this in the Core kind of limits on what we can do with the application concept. Different organizations might want to tweak this experience based on limitations they see in their enterprises and CRD seems like the best way to tweak those experiences. again my opinion might change as i  read through the full proposal, but it does seem like it should start as a CRD concept . the core should be minimal and extensible

Reply:
I think we have agreement that we will use a CRD